File: android/src/main/java/com/bradbumbalough/RCTTwilioChat/RCTTwilioChatClient.java
package com.bradbumbalough.RCTTwilioChat;

import android.support.annotation.Nullable;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.ReadableMap;

import com.facebook.react.modules.core.DeviceEventManagerModule;
import com.twilio.accessmanager.AccessManager;
import com.twilio.chat.Channel;
import com.twilio.chat.StatusListener;
import com.twilio.chat.CallbackListener;
import com.twilio.chat.ErrorInfo;
import com.twilio.chat.ChatClientListener;
import com.twilio.chat.ChannelListener;
import com.twilio.chat.ChatClient;
import com.twilio.chat.User;
import com.twilio.chat.Message;

import java.util.HashMap;
import java.util.Map;

import org.json.JSONObject;
import android.util.Log;


public class RCTTwilioChatClient extends ReactContextBaseJavaModule implements ChatClientListener {

    @Override
    public String getName() {
        return "TwilioChatClient";
    }

    // Constants

    @Override
    public Map<String, Object> getConstants() {
        final Map<String, Object> constants = new HashMap<>();

        Map<String, String> channelStatus = new HashMap<>();
        channelStatus.put("Invited", Channel.ChannelStatus.INVITED.toString());
        channelStatus.put("Joined",Channel.ChannelStatus.JOINED.toString());
        channelStatus.put("NotParticipating",Channel.ChannelStatus.NOT_PARTICIPATING.toString());
        constants.put("TCHChannelStatus", channelStatus);

        Map<String, String> channelSyncStatus = new HashMap<>();
        channelSyncStatus.put("None",Channel.SynchronizationStatus.NONE.toString());
        channelSyncStatus.put("Identifier",Channel.SynchronizationStatus.IDENTIFIER.toString());
        channelSyncStatus.put("Metadata",Channel.SynchronizationStatus.METADATA.toString());
        channelSyncStatus.put("All",Channel.SynchronizationStatus.ALL.toString());
        channelSyncStatus.put("Failed",Channel.SynchronizationStatus.FAILED.toString());
        constants.put("TCHChannelSynchronizationStatus", channelSyncStatus);

        Map<String, String> channelType = new HashMap<>();
        channelType.put("Public",Channel.ChannelType.PUBLIC.toString());
        channelType.put("Private",Channel.ChannelType.PRIVATE.toString());
        constants.put("TCHChannelType", channelType);

        Map<String, String> clientSyncStatus = new HashMap<>();
        clientSyncStatus.put("Started",ChatClient.SynchronizationStatus.STARTED.toString());
        clientSyncStatus.put("ChannelListCompleted",ChatClient.SynchronizationStatus.CHANNELS_COMPLETED.toString());
        clientSyncStatus.put("Completed",ChatClient.SynchronizationStatus.COMPLETED.toString());
        clientSyncStatus.put("Failed",ChatClient.SynchronizationStatus.FAILED.toString());
        constants.put("TCHClientSynchronizationStatus", clientSyncStatus);

        Map<String, String> clientSyncStrategy = new HashMap<>();
        clientSyncStrategy.put("All",ChatClient.SynchronizationStrategy.ALL.toString());
        clientSyncStrategy.put("ChannelsList",ChatClient.SynchronizationStrategy.CHANNELS_LIST.toString());
        constants.put("TCHClientSynchronizationStrategy", clientSyncStrategy);

        Map<String, String> channelOption = new HashMap<>();
        channelOption.put("FriendlyName", "friendlyName");
        channelOption.put("UniqueName", "uniqueName");
        channelOption.put("Type", "type");
        channelOption.put("Attributes", "attributes");
        constants.put("TCHChannelOption", channelOption);

        Map<String, String> connectionState = new HashMap<>();
        connectionState.put("Connecting", ChatClient.ConnectionState.CONNECTING.toString());
        connectionState.put("Connected", ChatClient.ConnectionState.CONNECTED.toString());
        connectionState.put("Disconnected", ChatClient.ConnectionState.DISCONNECTED.toString());
        connectionState.put("Denied", ChatClient.ConnectionState.DENIED.toString());
        connectionState.put("Error", ChatClient.ConnectionState.FATAL_ERROR.toString());
        constants.put("TCHClientConnectionState", connectionState);

        Map<String, Integer> logLevel = new HashMap<>();
        logLevel.put("Fatal", Log.ERROR);
        logLevel.put("Critical", Log.ERROR);
        logLevel.put("Warning", Log.WARN);
        logLevel.put("Info", Log.INFO);
        logLevel.put("Debug", Log.DEBUG);
        constants.put("TCHLogLevel", logLevel);

        Map<String, String> user = new HashMap<>();
        user.put("Attributes", User.UpdateReason.ATTRIBUTES.toString());
        user.put("FriendlyName", User.UpdateReason.FRIENDLY_NAME.toString());
        user.put("ReachabilityNotifiable", User.UpdateReason.REACHABILITY_NOTIFIABLE.toString());
        user.put("ReachabilityOnline", User.UpdateReason.REACHABILITY_ONLINE.toString());
        constants.put("TCHUserUpdate", user);

        return constants;
    }

    public ChatClient client = null;
    private ReactApplicationContext reactContext;

    private static RCTTwilioChatClient rctTwilioChatClient;

    public static RCTTwilioChatClient getInstance() {
        return rctTwilioChatClient;
    }

    public RCTTwilioChatClient(ReactApplicationContext reactContext) {
        super(reactContext);
        this.reactContext = reactContext;
        rctTwilioChatClient = this;
    }

    private void sendEvent(String eventName, @Nullable WritableMap params) {
        RCTTwilioChatClient tmp = RCTTwilioChatClient.getInstance();
        tmp.reactContext
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                .emit(eventName, params);
    }

    private void sendEvent(String eventName, @Nullable String body) {
        RCTTwilioChatClient tmp = RCTTwilioChatClient.getInstance();
        tmp.reactContext
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                .emit(eventName, body);
    }

    private StatusListener generateStatusListener(final Promise promise, final String errorCode, final String errorMessage) {
        return new StatusListener() {
            @Override
            public void onError(ErrorInfo errorInfo) {
                super.onError(errorInfo);
                promise.reject(errorCode, errorMessage + " Error Message: " + errorInfo.getMessage());
            }

            @Override
            public void onSuccess() {
                promise.resolve(true);
            }
        };
    }

    // Methods

    @ReactMethod
    public void createClient(String token, ReadableMap props, final Promise promise) {
        ChatClient.Properties.Builder builder = new ChatClient.Properties.Builder();
        if (props != null) {
            if (props.hasKey("region")) {
                builder.setRegion(props.getString("region"));
            }
        }

        ChatClient.create(reactContext, token, builder.createProperties(), new CallbackListener<ChatClient>() {
            @Override
            public void onError(ErrorInfo errorInfo) {
                super.onError(errorInfo);
                promise.reject("create-client-error", "Error occurred while attempting to create the client. Error Message: " + errorInfo.getMessage());
            }

            @Override
            public void onSuccess(ChatClient twilioChatClient) {
                RCTTwilioChatClient tmp = RCTTwilioChatClient.getInstance();
                tmp.client = twilioChatClient;
                tmp.client.setListener(tmp);
                promise.resolve(RCTConvert.ChatClient(tmp.client));
            }
        });
    }

    @ReactMethod
    public void updateToken(String token,final Promise promise) {
        RCTTwilioChatClient tmp = RCTTwilioChatClient.getInstance();
        tmp.client.updateToken(token, new StatusListener() {
            @Override
            public void onSuccess() {
              promise.resolve(true);
            }

            @Override
            public void onError(ErrorInfo errorInfo) {
              super.onError(errorInfo);
              promise.reject("update-client--token-error", "Error occurred while attempting to update the client token. Error Message: " + errorInfo.getMessage());
            }
        });
    }

    @ReactMethod
    public void user(Promise promise) {
        RCTTwilioChatClient tmp = RCTTwilioChatClient.getInstance();
        promise.resolve(RCTConvert.User(tmp.client.getUsers().getMyUser()));
    }

    @ReactMethod
    public void register(String token) {
        RCTTwilioChatClient tmp = RCTTwilioChatClient.getInstance();

        StatusListener listener = new StatusListener() {
            @Override
            public void onError(ErrorInfo errorInfo) {
                super.onError(errorInfo);
            }

            @Override
            public void onSuccess() {

            }
        };
        tmp.client.registerGCMToken(token, listener);
    }

    @ReactMethod
    public void unregister(String token) {
        RCTTwilioChatClient tmp = RCTTwilioChatClient.getInstance();

        StatusListener listener = new StatusListener() {
            @Override
            public void onError(ErrorInfo errorInfo) {
                super.onError(errorInfo);
            }

            @Override
            public void onSuccess() {

            }
        };
        tmp.client.unregisterGCMToken(token, listener);
    }

    @ReactMethod
    public void handleNotification(ReadableMap notification) {
        RCTTwilioChatClient tmp = RCTTwilioChatClient.getInstance();
        HashMap map = RCTConvert.readableMapToHashMap(notification);
//        TODO
//        tmp.client.handleNotification(map);
    }

    @ReactMethod
    public void shutdown() {
        RCTTwilioChatClient tmp = RCTTwilioChatClient.getInstance();
        tmp.client.shutdown();
    }

    @ReactMethod
    public void setLogLevel(Integer logLevel) {
        ChatClient.setLogLevel(logLevel);
    }

    // UserInfo methods

    @ReactMethod
    public void setFriendlyName(String friendlyName, final Promise promise) {
        RCTTwilioChatClient tmp = RCTTwilioChatClient.getInstance();

        StatusListener listener = new StatusListener() {
            @Override
            public void onError(ErrorInfo errorInfo) {
                super.onError(errorInfo);
                promise.reject("set-friendly-name-error", "Error occured while attempting to set friendly name for the user.");
            }

            @Override
            public void onSuccess() {
                promise.resolve(true);
            }
        };

        tmp.client.getUsers().getMyUser().setFriendlyName(friendlyName, listener);
    }

    @ReactMethod
    public void setAttributes(ReadableMap attributes, final Promise promise) {
        RCTTwilioChatClient tmp = RCTTwilioChatClient.getInstance();
        JSONObject json = RCTConvert.readableMapToJson(attributes);

        StatusListener listener = new StatusListener() {
            @Override
            public void onError(ErrorInfo errorInfo) {
                super.onError(errorInfo);
                promise.reject("set-attributes-error", "Error occured while attempting to set attributes for the user.");
            }

            @Override
            public void onSuccess() {
                promise.resolve(true);
            }
        };

        tmp.client.getUsers().getMyUser().setAttributes(json, listener);
    }

    // Listeners

    @Override
    public void onConnectionStateChange(ChatClient.ConnectionState state) {
        sendEvent("chatClient:connectionStateChanged", state.toString());
    }

    @Override
    public void onChannelAdded(final Channel channel) {
        sendEvent("chatClient:channelAdded", RCTConvert.Channel(channel));
    }

    @Override
    public void onChannelJoined(Channel channel) {
        sendEvent("chatClient:channelJoined", RCTConvert.Channel(channel));
    }

    @Override
    public void onChannelInvited(Channel channel) {
        sendEvent("chatClient:channelInvited", RCTConvert.Channel(channel));
    }

    @Override
    public void onChannelUpdated(Channel channel, Channel.UpdateReason reason) {
        sendEvent("chatClient:channel:updated", RCTConvert.Channel(channel));
    }

    @Override
    public void onChannelDeleted(Channel channel){
        sendEvent("chatClient:channelDeleted", RCTConvert.Channel(channel));
    }

    @Override
    public void onChannelSynchronizationChange(Channel channel) {
        WritableMap map = Arguments.createMap();
        map.putString("channelSid",channel.getSid());
        map.putString("status", channel.getSynchronizationStatus().toString());
        sendEvent("chatClient:channel:synchronizationStatusChanged", map);
    }

    @Override
    public void onClientSynchronization(ChatClient.SynchronizationStatus synchronizationStatus) {
        sendEvent("chatClient:synchronizationStatusUpdated", synchronizationStatus.toString());
    }

    @Override
    public void onUserUpdated(User user, User.UpdateReason updateReason) {
        WritableMap map = Arguments.createMap();
        map.putString("updated", updateReason.toString());
        map.putMap("user", RCTConvert.User(user));
        sendEvent("chatClient:userUpdated", map);
    }

    @Override
    public void onUserUnsubscribed(User user) {
        WritableMap map = Arguments.createMap();
        map.putMap("user", RCTConvert.User(user));
        sendEvent("chatClient:userUnsubscribed", map);
    }

    @Override
    public void onUserSubscribed(User user) {
        WritableMap map = Arguments.createMap();
        map.putMap("user", RCTConvert.User(user));
        sendEvent("chatClient:userUnsubscribed", map);
    }

    @Override
    public void onError(ErrorInfo errorInfo) {
        WritableMap map = Arguments.createMap();
        map.putString("error",errorInfo.getMessage());
        map.putString("userInfo", errorInfo.toString());
        sendEvent("chatClient:errorReceived", map);
    }

    @Override
    public void onNotificationSubscribed() {
        sendEvent("chatClient:toastSubscribed", "");
    }

    @Override
    public void onNotification(String channelSid, String messageSid) {
        WritableMap map = Arguments.createMap();
        map.putString("channelSid", channelSid);
        map.putString("messageSid", messageSid);

        sendEvent("chatClient:toastReceivedOnChannel", map);
    }

    @Override
    public void onNotificationFailed(ErrorInfo errorInfo) {
        WritableMap map = Arguments.createMap();
        map.putString("error",errorInfo.getMessage());
        map.putString("userInfo", errorInfo.toString());

        sendEvent("chatClient:toastFailed", map);
    }
}

File: ios/RCTTwilioChat/RCTTwilioChatClient.m
//
//  RCTTwilioChatClient.m
//  TwilioIPExample
//
//  Created by Brad Bumbalough on 5/31/16.
//  Copyright © 2016 Facebook. All rights reserved.
//

#import "RCTTwilioChatClient.h"
#import "RCTTwilioAccessManager.h"
#import <React/RCTEventDispatcher.h>
#import "RCTConvert+TwilioChatClient.h"
#import <React/RCTUtils.h>

@interface RCTTwilioChatClient() <TwilioChatClientDelegate>
@end

@implementation RCTTwilioChatClient

@synthesize bridge = _bridge;
@synthesize client;

#pragma mark Singleton Methods

+ (id)sharedManager {
  static RCTTwilioChatClient *sharedMyManager = nil;
  static dispatch_once_t onceToken;
  dispatch_once(&onceToken, ^{
    sharedMyManager = [[self alloc] init];
  });
  return sharedMyManager;
}

RCT_EXPORT_MODULE()

RCT_REMAP_METHOD(createClient, token:(NSString*)token properties:(NSDictionary *)properties resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    TwilioChatClientProperties *props = nil;
    if (properties.count > 0) {
      props = [[TwilioChatClientProperties alloc] init];
      if (properties[@"region"]) {
        props.region = [RCTConvert NSString:properties[@"region"]];
      }
    }
    [TwilioChatClient chatClientWithToken:token properties:props delegate:self completion:^(TCHResult *result, TwilioChatClient *chatClient) {
      if (!result.isSuccessful) {
        reject(@"create-client-error", @"Error occurred while attempting to create the [client. Error Message:", result.error);
        return;
      }
      RCTTwilioChatClient *_client = [RCTTwilioChatClient sharedManager];
      _client.client = chatClient;
      resolve([RCTConvert TwilioChatClient:_client.client]);
    }];
}

RCT_EXPORT_METHOD(updateToken:(NSString *)token) {
    RCTTwilioChatClient *_client = [RCTTwilioChatClient sharedManager];
    [[_client client] updateToken:token completion:^(TCHResult *result) {
      if (result.isSuccessful) {
        NSLog(@"Client token updated");
      }
      else {
        NSLog(@"Client token update failed: %@", result.error.localizedDescription);
      }
    }];
}

RCT_REMAP_METHOD(user, userInfo_resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  RCTTwilioChatClient *_client = [RCTTwilioChatClient sharedManager];
  resolve([RCTConvert TCHUser:_client.client.user]);
}

RCT_EXPORT_METHOD(synchronizationStatus:(RCTResponseSenderBlock)callback) {
  RCTTwilioChatClient *_client = [RCTTwilioChatClient sharedManager];
  callback(@[@(_client.client.synchronizationStatus)]);
}

RCT_EXPORT_METHOD(version:(RCTResponseSenderBlock)callback) {
  callback(@[client.version]);
}

RCT_EXPORT_METHOD(register:(NSString *)token) {
  RCTTwilioChatClient *_client = [RCTTwilioChatClient sharedManager];
  [[_client client] registerWithNotificationToken:[RCTConvert dataWithHexString:token] completion:^(TCHResult *result) {
    if (result.isSuccessful) {
      NSLog(@"Notification token registered");
    }
    else {
      NSLog(@"Notification token registration failed: %@", result.error.localizedDescription);
    }
  }];
}

RCT_EXPORT_METHOD(unregister:(NSString *)token) {
  RCTTwilioChatClient *_client = [RCTTwilioChatClient sharedManager];
  [[_client client] deregisterWithNotificationToken:[RCTConvert dataWithHexString:token] completion:^(TCHResult *result) {
    if (result.isSuccessful) {
      NSLog(@"Notification token deregistered");
    }
    else {
      NSLog(@"Notification token deregistration failed: %@", result.error.localizedDescription);
    }
  }];
}

RCT_EXPORT_METHOD(handleNotification:(NSDictionary *)notification) {
  RCTTwilioChatClient *_client = [RCTTwilioChatClient sharedManager];
  [[_client client] handleNotification:notification completion:^(TCHResult *result) {
    if (result.isSuccessful) {
      NSLog(@"Notification handled successfully");
    }
    else {
      NSLog(@"Notification handling failed: %@", result.error.localizedDescription);
    }
  }];
}

RCT_EXPORT_METHOD(shutdown) {
  RCTTwilioChatClient *_client = [RCTTwilioChatClient sharedManager];
  [[_client client] shutdown];
}

RCT_EXPORT_METHOD(setLogLevel:(TCHLogLevel)logLevel) {
  [TwilioChatClient setLogLevel:logLevel];
}

RCT_EXPORT_METHOD(logLevel:(TCHLogLevel)logLevel callback:(RCTResponseSenderBlock)callback) {
  callback(@[@(TwilioChatClient.logLevel)]);
}

RCT_REMAP_METHOD(setFriendlyName, friendlyName:(NSString *)friendlyName friendlyName_resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  RCTTwilioChatClient *_client = [RCTTwilioChatClient sharedManager];
  [[[_client client] user] setFriendlyName:friendlyName completion:^(TCHResult *result) {
    if (result.isSuccessful) {
      resolve(@[@TRUE]);
    }
    else {
      reject(@"set-friendly-name-error", @"Error occured while attempting to set friendly name for the user.", result.error);
    }
  }];
}

RCT_REMAP_METHOD(setAttributes, attributes:(NSDictionary *)attributes attributes_resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  RCTTwilioChatClient *_client = [RCTTwilioChatClient sharedManager];
  [[[_client client] user] setAttributes:attributes completion:^(TCHResult *result) {
    if (result.isSuccessful) {
      resolve(@[@TRUE]);
    }
    else {
      reject(@"set-attributes-error", @"Error occured while attempting to set attributes for the user.", result.error);
    }
  }];
}

#pragma mark Twilio IP Messaging Client Delegates

- (void)chatClient:(TwilioChatClient *)client connectionStateChanged:(TCHClientConnectionState)state {
    [self.bridge.eventDispatcher sendAppEventWithName:@"chatClient:connectionStateChanged"
                                                 body:@(state)];
}

- (void)chatClient:(TwilioChatClient *)client synchronizationStatusUpdated:(TCHClientSynchronizationStatus)status {
  [self.bridge.eventDispatcher sendAppEventWithName:@"chatClient:synchronizationStatusUpdated"
                                               body:@(status)];
}

- (void)chatClient:(TwilioChatClient *)client channelAdded:(TCHChannel *)channel {
  [self.bridge.eventDispatcher sendAppEventWithName:@"chatClient:channelAdded"
                                               body: [RCTConvert TCHChannel:channel]];
}

- (void)chatClient:(TwilioChatClient *)client channel:(TCHChannel *)channel updated:(TCHChannelUpdate)updated {
  [self.bridge.eventDispatcher sendAppEventWithName:@"chatClient:channel:updated"
                                               body:[RCTConvert TCHChannel:channel]];
}

- (void)chatClient:(TwilioChatClient *)client channelDeleted:(TCHChannel *)channel {
  [self.bridge.eventDispatcher sendAppEventWithName:@"chatClient:channelDeleted"
                                               body: [RCTConvert TCHChannel:channel]];
}

- (void)chatClient:(TwilioChatClient *)client channel:(TCHChannel *)channel synchronizationStatusChanged:(TCHChannelSynchronizationStatus)status {
  NSString *channelSid;
  if (channel) {
    channelSid = channel.sid;
  }
  else {
    channelSid = RCTNullIfNil(nil);
  }
  [self.bridge.eventDispatcher sendAppEventWithName:@"chatClient:channel:synchronizationStatusChanged"
                                               body: @{
                                                       @"channelSid": channelSid,
                                                       @"status": @(status)
                                                       }];
}

- (void)chatClient:(TwilioChatClient *)client channel:(TCHChannel *)channel memberJoined:(TCHMember *)member {
  [self.bridge.eventDispatcher sendAppEventWithName:@"chatClient:channel:memberJoined"
                                               body: @{
                                                       @"channelSid": channel.sid,
                                                       @"member": [RCTConvert TCHMember:member]
                                                       }];
}

- (void)chatClient:(TwilioChatClient *)client channel:(TCHChannel *)channel memberLeft:(TCHMember *)member {
  [self.bridge.eventDispatcher sendAppEventWithName:@"chatClient:channel:memberLeft"
                                               body: @{
                                                       @"channelSid": channel.sid,
                                                       @"member": [RCTConvert TCHMember:member]
                                                       }];
}

- (void)chatClient:(TwilioChatClient *)client channel:(TCHChannel *)channel messageAdded:(TCHMessage *)message {
  [self.bridge.eventDispatcher sendAppEventWithName:@"chatClient:channel:messageAdded"
                                               body: @{
                                                       @"channelSid": channel.sid,
                                                       @"message": [RCTConvert TCHMessage:message]
                                                       }];
}

- (void)chatClient:(TwilioChatClient *)client channel:(TCHChannel *)channel message:(TCHMessage *)message updated:(TCHMessageUpdate)updated {
  [self.bridge.eventDispatcher sendAppEventWithName:@"chatClient:channel:messageChanged"
                                               body: @{
                                                       @"channelSid": channel.sid,
                                                       @"message": [RCTConvert TCHMessage:message]
                                                       }];
}

- (void)chatClient:(TwilioChatClient *)client channel:(TCHChannel *)channel messageDeleted:(TCHMessage *)message {
  [self.bridge.eventDispatcher sendAppEventWithName:@"chatClient:channel:messageDeleted"
                                               body: @{
                                                       @"channelSid": channel.sid,
                                                       @"message": [RCTConvert TCHMessage:message]
                                                       }];
}

- (void)chatClient:(TwilioChatClient *)client errorReceived:(TCHError *)error {
  [self.bridge.eventDispatcher sendAppEventWithName:@"chatClient:errorReceived"
                                               body:@{@"error": [error localizedDescription],
                                                      @"userInfo": [error userInfo]
                                                      }];
}

- (void)chatClient:(TwilioChatClient *)client typingStartedOnChannel:(TCHChannel *)channel member:(TCHMember *)member {
  [self.bridge.eventDispatcher sendAppEventWithName:@"chatClient:typingStartedOnChannel"
                                               body: @{
                                                       @"channelSid": channel.sid,
                                                       @"member": [RCTConvert TCHMember:member]
                                                       }];
}

- (void)chatClient:(TwilioChatClient *)client typingEndedOnChannel:(TCHChannel *)channel member:(TCHMember *)member {
  [self.bridge.eventDispatcher sendAppEventWithName:@"chatClient:typingEndedOnChannel"
                                               body: @{
                                                       @"channelSid": channel.sid,
                                                       @"member": [RCTConvert TCHMember:member]
                                                       }];
}

- (void)chatClientToastSubscribed:(TwilioChatClient *)client {
  [self.bridge.eventDispatcher sendAppEventWithName:@"chatClient:toastSubscribed"
                                               body: RCTNullIfNil(nil)];
}

- (void)chatClient:(TwilioChatClient *)client toastReceivedOnChannel:(TCHChannel *)channel message:(TCHMessage *)message {
    [self.bridge.eventDispatcher sendAppEventWithName:@"chatClient:toastReceivedOnChannel"
                                               body: @{
                                                       @"channelSid": channel.sid,
                                                       @"messageSid": message.sid
                                                       }];
}

- (void)chatClient:(TwilioChatClient *)client toastRegistrationFailedWithError:(TCHError *)error {
  [self.bridge.eventDispatcher sendAppEventWithName:@"chatClient:toastFailed"
                                               body:@{@"error": [error localizedDescription],
                                                      @"userInfo": @""
                                                      }];
}

- (void)chatClient:(TwilioChatClient *)client user:(TCHUser *)user updated:(TCHUserUpdate)updated {
  [self.bridge.eventDispatcher sendAppEventWithName:@"chatClient:userUpdated"
                                               body: @{@"updated": @(updated),
                                                       @"user": [RCTConvert TCHUser:user]
                                                       }];
}

- (void)chatClient:(TwilioChatClient *)client channel:(TCHChannel *)channel member:(TCHMember *)member updated:(TCHMemberUpdate)updated {
  [member userDescriptorWithCompletion:^(TCHResult *result, TCHUserDescriptor *user) {
    [self.bridge.eventDispatcher sendAppEventWithName:@"chatClient:channel:member:userUpdated"
                                                 body: @{@"channelSid": channel.sid,
                                                         @"updated": @(updated),
                                                         @"userDescriptor": [RCTConvert TCHUserDescriptor:user]
                                                         }];
  }];

}

#pragma mark Enums

- (NSDictionary *)constantsToExport
{
  return @{ @"Constants": @{
                @"TCHClientSynchronizationStatus": @{
                    @"Started" : @(TCHClientSynchronizationStatusStarted),
                    @"ChannelsListCompleted" : @(TCHClientSynchronizationStatusChannelsListCompleted),
                    @"Completed" : @(TCHClientSynchronizationStatusCompleted),
                    @"Failed" : @(TCHClientSynchronizationStatusFailed)
                    },
                @"TCHChannelSynchronizationStatus": @{
                    @"None" : @(TCHChannelSynchronizationStatusNone),
                    @"Identifier" : @(TCHChannelSynchronizationStatusIdentifier),
                    @"Metadata" : @(TCHChannelSynchronizationStatusMetadata),
                    @"All" : @(TCHChannelSynchronizationStatusAll),
                    @"Failed" : @(TCHChannelSynchronizationStatusFailed)
                    },
                @"TCHChannelStatus": @{
                    @"Invited": @(TCHChannelStatusInvited),
                    @"Joined": @(TCHChannelStatusJoined),
                    @"NotParticipating": @(TCHChannelStatusNotParticipating)
                    },
                @"TCHChannelType": @{
                    @"Public": @(TCHChannelTypePublic),
                    @"Private": @(TCHChannelTypePrivate)
                    },
                @"TCHUserInfoUpdate": @{
                    @"FriendlyName": @(TCHUserUpdateFriendlyName),
                    @"Attributes": @(TCHUserUpdateAttributes),
                    @"ReachabilityOnline": @(TCHUserUpdateReachabilityOnline),
                    @"ReachabilityNotifiable": @(TCHUserUpdateReachabilityNotifiable)
                    },
                @"TCHLogLevel": @{
                    @"Fatal" : @(TCHLogLevelFatal),
                    @"Critical" : @(TCHLogLevelCritical),
                    @"Warning" : @(TCHLogLevelWarning),
                    @"Info" : @(TCHLogLevelInfo),
                    @"Debug" : @(TCHLogLevelDebug)
                    },
                @"TCHChannelOption": @{
                        @"FriendlyName" : TCHChannelOptionFriendlyName,
                        @"UniqueName" : TCHChannelOptionUniqueName,
                        @"Type" : TCHChannelOptionType,
                        @"Attributes" : TCHChannelOptionAttributes
                        },
                @"TCHClientConnectionState": @{
                        @"Unknown" : @(TCHClientConnectionStateUnknown),
                        @"Disconnected" : @(TCHClientConnectionStateDisconnected),
                        @"Connected" : @(TCHClientConnectionStateConnected),
                        @"Connecting" : @(TCHClientConnectionStateConnecting),
                        @"Denied" : @(TCHClientConnectionStateDenied),
                        @"Error" : @(TCHClientConnectionStateError)
                        }
                }
            };
};


@end
